Proposal: Ordered Collection Diffing

numist opened the pull request:
<!-- What's in this pull request? -->
This PR is an initial implementation of the [Ordered Collection Diffing proposal](https://github.com/apple/swift-evolution/pull/968), which was [pitched on the forums](https://forums.swift.org/t/ordered-collection-diffing/18933) in December.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

numist committed:
Initial import of diffing package into standard library

numist committed:
Adapt the last of the Darwin- and XCTest-specific code in the diffing tests

Azoy commented:
Nit: stdlib uses 2 space tabs

numist committed:
Match indentation style

numist committed:
s/shortestEditScript/difference/g per proposal feedback

numist committed:
Apply public documentation and type modifications from proposal acceptance

numist committed:
Include renamed CollectionDifference.swift

numist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,332 @@
+  }
+
+  // The public initializer calls this function to ensure that its parameter
+  // meets the conditions set in its documentation.
```
This does both; the top-order of complexity comes from sorting, which is required to check the inputs for validity as well as provide the stable enumeration order and `insertions`/`removals` properties promised by the API.

dabrahams replied:
Thanks!  The high-order bit of my request, though, is “please fully document even the non-public decls.”

numist replied:
Understood! Please see 4552291

numist committed:
Header docs for internal methods validateChanges(_:) and init(validatedChanges:)

numist committed:
Only instanciate one test suite--this was confusing the test framework, causing some strange crashes

DevAndArtist reviewed (changes requested):
@numist several lines of this patch are too long. Stdlib uses 2 spaces tabs and has a maximum line width set to 80 characters.

DevAndArtist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+  ///   3. All associations between insertions and removals are symmetric
+  ///
+  /// Complexity: O(`changes.count`)
+  private static func validateChanges<C>(_ changes : C) -> Bool where C:Collection, C.Element == Change {
```
```suggestion
private static func validateChanges<C>(
  _ changes: C
) -> Bool where C: Collection, C.Element == Change {
```

DevAndArtist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+  ///
+  /// - Complexity: O(*n* * log(*n*)), where *n* is the length of the
+  ///   parameter.
+  public init?<C>(_ c: C) where C:Collection, C.Element == Change {
```
```suggestion
  public init?<C>(_ c: C) where C: Collection, C.Element == Change {
```

DevAndArtist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+  ///
+  /// - Complexity: O(*n* * log(*n*)), where *n* is the length of the
+  ///   parameter.
+  init<C>(validatedChanges c: C) where C:Collection, C.Element == Change {
```
```suggestion
  init<C>(validatedChanges c: C) where C: Collection, C.Element == Change {
```

DevAndArtist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+
+  // Opaque index type is isomorphic to Int
+  public struct Index: Comparable, Hashable {
+    public static func < (lhs: CollectionDifference<ChangeElement>.Index, rhs: CollectionDifference<ChangeElement>.Index) -> Bool {
```
```suggestion
    public static func < (
      lhs: CollectionDifference<ChangeElement>.Index, 
      rhs: CollectionDifference<ChangeElement>.Index
    ) -> Bool {
```

DevAndArtist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+    return Index(removals.count + insertions.count)
+  }
+  
+  public func index(after index: CollectionDifference<ChangeElement>.Index) -> CollectionDifference<ChangeElement>.Index {
```
```suggestion
  public func index(
    after index: CollectionDifference<ChangeElement>.Index
  ) -> CollectionDifference<ChangeElement>.Index {
```

DevAndArtist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+    return Index(index.i + 1)
+  }
+  
+  public subscript(position: CollectionDifference<ChangeElement>.Index) -> Element {
```
I don't see the line width in the browser, but you should get an idea by now of what should be fixed.

lorentey commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+      return res.filter { (_, v) -> Bool in v != .none }
+    }()
+
+    return CollectionDifference.init(validatedChanges:map({ (c: CollectionDifference<ChangeElement>.Change) -> CollectionDifference<ChangeElement>.Change in
```
The `Change` type can be used unqualified within `CollectionDifference`, which should help making this a lot more readable:

```suggestion
    return CollectionDifference(validatedChanges: map { (change: Change) -> Change in
```

I'm pretty sure we don't need provide the closure's argument/return types explicitly, either.

lorentey commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+}
+
+extension CollectionDifference.Change: Codable where ChangeElement: Codable {
+  private enum CodingKeys: String, CodingKey {
```
```suggestion
  private enum _CodingKeys: String, CodingKey {
```

(Foundation isn't pedantic about using the underscores, but the stdlib core is, and AFAIK the type name is not required to literally be `CodingKeys`.)

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+// MARK: Diff application to RangeReplaceableCollection
+
+extension RangeReplaceableCollection {
+  private static func fastApplicationEnumeration(
```
This name needs an underscore, and it might work better as a method on `CollectionDifference`.
Additionally, the name of the function argument should be expanded. One random suggestion:

```suggestion
extension CollectionDifference {
  fileprivate func _fastEnumeratedApply(_ consume: (Change) -> Void) {...}
}
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+    let source = CountingIndexCollection(other)
+    let target = CountingIndexCollection(self)
+    let result = CollectionChanges(from: source, to: target, by: areEquivalent)
+      for c in result {
```
c ⟹ change

Nit: There is an extra level of indentation here.

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+
+    let source = CountingIndexCollection(other)
+    let target = CountingIndexCollection(self)
+    let result = CollectionChanges(from: source, to: target, by: areEquivalent)
```
Is `result` really a good name here? Perhaps it would be okay to just generate the changes directly in the `for` statement:

```swift
    for change in CollectionChanges(from: source, to: target, by: areEquivalent) {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+extension BidirectionalCollection {
+  /// Returns a pair of subsequences containing the initial elements that
+  /// `self` and `other` have in common.
+  fileprivate func commonPrefix<Other : BidirectionalCollection>(
```
`commonPrefix` ⟹ `_commonPrefix`

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+  /// Creates a collection of changes from a difference path.
+  init(
+    pathStorage: [Endpoint], pathStartIndex: Int
+  ) {
```
We need the explicit `[file]private` modifier:

```swift
  private init(
    pathStorage: [Endpoint], 
    pathStartIndex: Int
  ) {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+/// - Note: `CollectionChanges` holds a reference to state used to run the
+///     difference algorithm, which can be exponentially larger than the
+///     changes themselves.
+fileprivate struct CollectionChanges<
```
We want an underscore in the type name here. 

```suggestion
fileprivate struct _CollectionChanges<
```

Luckily, as long as we have that, the member names can be left without an underscore prefix.

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+
+extension CollectionChanges : CustomStringConvertible {
+  var description: String {
+    return String(describing: Array(self))
```
We have a dedicated function for this in the stdlib:

```suggestion
    return _makeCollectionDescription()
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+fileprivate struct CollectionChanges<
+  SourceIndex : Comparable, TargetIndex : Comparable
+> {
+  typealias Endpoint = (x: SourceIndex, y: TargetIndex)
```
We'll need an explicit access level for the typealias, too:
```suggestion
  private typealias Endpoint = (x: SourceIndex, y: TargetIndex)
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+extension CollectionChanges {
+  /// A range of elements removed from the source, inserted in the target, or
+  /// that the source and target have in common.
+  enum Element {
```
```suggestion
  private enum Element {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+}
+
+extension CollectionChanges : RandomAccessCollection {
+  typealias Index = Int
```
```suggestion
  private typealias Index = Int
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+extension CollectionChanges : RandomAccessCollection {
+  typealias Index = Int
+
+  var startIndex: Index {
```
```suggestion
  private var startIndex: Index {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+    return 0
+  }
+
+  var endIndex: Index {
```
```suggestion
  private var endIndex: Index {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+    return Swift.max(0, pathStorage.endIndex - pathStartIndex - 1)
+  }
+
+  func index(after i: Index) -> Index {
```
```suggestion
  private func index(after i: Index) -> Index {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+    return i + 1
+  }
+
+  subscript(position: Index) -> Element {
```
```suggestion
  private subscript(position: Index) -> Element {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+  ///   *d* is the minimal number of inserted and removed elements.
+  /// - Space: O(*d* * *d*), where *d* is the minimal number of inserted and
+  ///   removed elements.
+  init<Source : BidirectionalCollection, Target : BidirectionalCollection>(
```
```suggestion
  private init<Source : BidirectionalCollection, Target : BidirectionalCollection>(
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+  ///   *d* is the minimal number of inserted and removed elements.
+  /// - Space: O(*d*²), where *d* is the minimal number of inserted and
+  ///   removed elements.
+  mutating func formChanges<
```
```suggestion
  private mutating func formChanges<
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+  /// Creates an instance, taking the capacity of `storage` for itself.
+  ///
+  /// - Postcondition: `storage` is empty.
+  init(consuming storage: inout [Endpoint]) {
```
```suggestion
  private init(consuming storage: inout [Endpoint]) {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+
+  /// Returns the endpoint of the search frontier for iteration `d` on
+  /// diagonal `k`.
+  subscript(d: Int, k: Int) -> Endpoint {
```
```suggestion
  private subscript(d: Int, k: Int) -> Endpoint {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+
+  /// Adds endpoints initialized to `repeatedValue` for the search frontier of
+  /// the next iteration.
+  mutating func appendFrontier(repeating repeatedValue: Endpoint) {
```
```suggestion
  private mutating func appendFrontier(repeating repeatedValue: Endpoint) {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+}
+
+/// The search paths being explored.
+fileprivate struct SearchState<
```
```suggestion
fileprivate struct _SearchState<
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+  /// Removes and returns `CollectionChanges`, leaving `SearchState` empty.
+  ///
+  /// - Precondition: There is at least one difference between `a` and `b`
+  mutating func removeCollectionChanges<
```
```suggestion
  private mutating func removeCollectionChanges<
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+}
+
+/// An index that counts its offset from the start of its collection.
+struct CountingIndex<Base : Comparable> : Equatable {
```
This is implicitly internal, so it should have an underscore. (Maybe it could also be made private.)

```suggestion
internal struct _CountingIndex<Base : Comparable> : Equatable {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+}
+
+extension CountingIndex : Comparable {
+  static func <(lhs: CountingIndex, rhs: CountingIndex) -> Bool {
```
```suggestion
  internal static func <(lhs: CountingIndex, rhs: CountingIndex) -> Bool {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+///   // Prints "2"
+///
+/// - Note: The offset of `endIndex` is `nil`
+struct CountingIndexCollection<Base : BidirectionalCollection> {
```
```suggestion
internal struct _CountingIndexCollection<Base : BidirectionalCollection> {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+///
+/// - Note: The offset of `endIndex` is `nil`
+struct CountingIndexCollection<Base : BidirectionalCollection> {
+  let base: Base
```
```suggestion
  internal let base: Base
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+struct CountingIndexCollection<Base : BidirectionalCollection> {
+  let base: Base
+
+  init(_ base: Base) {
```
```suggestion
  internal init(_ base: Base) {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+  }
+}
+
+extension CountingIndexCollection : BidirectionalCollection {
```
We need to add explicit `internal` access levels throughout this extension.

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+/// Returns the nth [triangular number].
+///
+/// [triangular number]: https://en.wikipedia.org/wiki/Triangular_number
+fileprivate func triangularNumber(_ n: Int) -> Int {
```
```suggestion
fileprivate func _triangularNumber(_ n: Int) -> Int {
```

lorentey commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,782 @@
+///   ])
+///
+/// [lower triangular matrix]: http://en.wikipedia.org/wiki/Triangular_matrix
+fileprivate struct LowerTriangularMatrix<Element> {
```
```suggestion
fileprivate struct _LowerTriangularMatrix<Element> {
```

lorentey replied:
We also need to add explicit `private`/`fileprivate` modifiers for all members, including the ones defined in the extensions below. (But we don't need more underscores.)

lorentey reviewed (changes requested):
I had suggested a number of small(ish) adjustments -- mostly to follow the stdlib coding guidelines:

1. The fully qualified names of internal/private types (and members) must have at least one underscore-prefixed component. For initializers, we put the underscore in the label of the first parameter.
2. All declarations must have an explicit access level in the stdlib. (This is often redundant, but it helps a lot with reviewing what gets into the public API/ABI.)
3. When possible, we prefer to have meaningful names for generic type parameters, function parameters and local variables. (I think `x`, `y` etc. works fine in the core diffing implementation, though.)

lorentey committed:
[stdlib] Update diffing code to match stdlib coding conventions

- Add an underscore to private/internal symbols
- Make access levels explicit, even when they’re implied from context
- Conform to max line length of 80 characters
- Conformances declared on separate extensions that implement them
- Arrange member declarations so that stored properties appear first
- Expand single-letter function and type parameter names where there is an obvious name that’s more descriptive
- RangeReplaceableCollection.fastApplicationEnumeration → CollectionDifference._fastEnumeratedApply

lorentey committed:
[stdlib] Set availability of diffing APIs

numist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+  /// each `remove` refers to the offset of its `element` in the original
+  /// state. Non-`nil` values of `associatedWith` refer to the offset of the
+  /// complementary change.
+  public enum Change {
```
Yup, `@_frozen` is appropriate here.

numist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+  public typealias Element = CollectionDifference<ChangeElement>.Change
+
+  // Opaque index type is isomorphic to Int
+  public struct Index: Comparable, Hashable {
```
At the moment I think we should maintain the flexibility to change the internal representation in the future as adoption may guide us towards a more efficient solution.

DevAndArtist committed:
Wrap `<` to fit in 80

Suggested by @DevAndArtist

Co-Authored-By: numist <github@numist.net>

numist committed:
Merge branch 'numist/diffing' into numist/diffing

numist committed:
Merge pull request #1 from lorentey/numist/diffing

Preparations for landing numist/swift

numist commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,353 @@
+          res[element] = .some(r.offset)
+        }
+      }
+      return res.filter { (_, v) -> Bool in v != .none }
```
I'm sure there's a way to do this without filtering. Until now I was trying to maintain a simple implementation to simplify debugging, but now that the tests all pass (and provide sufficient coverage) it's safe to optimize 🙂

DevAndArtist commented:
I hope it won't get shipped with `@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)` by accident. 😅

moiseev commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,378 @@
+  public init?<Changes: Collection>(
+    _ changes: Changes
+  ) where Changes.Element == Change {
+    if !CollectionDifference<ChangeElement>._validateChanges(changes) {
```
FWIW: there's an idiomatic way to spell `if !` called `guard`. Some people prefer it for preconditions.

numist replied:
Oops, this actually used to be a `guard` but the code got refactored a number of times >_<

moiseev commented on stdlib/public/core/CollectionDifference.swift:
```diff
@@ -0,0 +1,378 @@
+}
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) // FIXME(availability-5.1)
+extension CollectionDifference.Index { // Comparable
```
Can't the conformance be removed from the declaration above, and uncommented here?

moiseev commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,805 @@
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) // FIXME(availability-5.1)
+extension CollectionDifference {
+  fileprivate func _fastEnumeratedApply(
```
At some point only `internal` and `public` declarations were allowed in the stdlib. I wonder what happened to that rule. @lorentey, @milseman do you know?

milseman replied:
I believe that private and fileprivate can work (but not as usableFromInline) in a world where we're not sil-serialize-all

lorentey replied:
We started to occasionally use private in the stdlib last year. It's still a bit unusual, and the fileprivate/private mess has annoying interactions with our rule for explicit access levels. But as far as I know, the technical issues we had before have all been resolved now.

moiseev commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,805 @@
+    func append(
+      into target: inout Self,
+      contentsOf source: Self,
+      from index: inout Self.Index, count: Int) {
```
The convention is to put a closing paren and the opening curly on the new line unindented after the parameter list.

moiseev commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,805 @@
+        enumeratedOriginals += origCount
+        enumeratedInserts += 1
+      }
+      assert(enumeratedOriginals <= self.count)
```
`assert` is rarely the right choice for the stdlib code. Consider using `_precondition`, `_debugPrecondition`, or `_internalInvariant` as described [here](https://github.com/apple/swift/blob/master/docs/StandardLibraryProgrammersManual.md#precondition).

moiseev commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,805 @@
+    by areEquivalent: (Element, C.Element) -> Bool
+  ) -> CollectionDifference<Element>
+  where C.Element == Self.Element
+  {
```
nitpick: I don't think we leave opening curlys on new line.

moiseev commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,805 @@
+  @available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) // FIXME(availability-5.1)
+  public func difference<C: BidirectionalCollection>(
+    from other: C
+  ) -> CollectionDifference<Element> where C.Element == Self.Element {
```
nitpick: in the extension above a `where` clause got its own line.

moiseev commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,805 @@
+  }
+
+  fileprivate subscript(position: Index) -> Element {
+    precondition((startIndex..<endIndex).contains(position))
```
`s/precondition/_precondition` perhaps?

numist replied:
I think `_internalInvariant` is actually more appropriate here since this only provides early detection of a code defect that would fail further along regardless. I'll go through all the calls to `assert` and update as appropriate.

moiseev commented on stdlib/public/core/Diffing.swift:
```diff
@@ -0,0 +1,805 @@
+    var (x, y) = (x, y)
+    let (n, m) = (a.endIndex, b.endIndex)
+
+    var v = _SearchState<Source.Index, Target.Index>(consuming: &pathStorage)
```
Vrbl nms R nt vry rdbl.

numist replied:
Sorry, too used to writing pure C—vowels are expensive!

lorentey replied:
I think the terse names are justified here since the implementation follows the paper, and using the same notation makes the code easier to follow.

moiseev commented on test/stdlib/Diffing.swift:
```diff
@@ -0,0 +1,667 @@
+
+  for (source, target, expected, line) in expectedChanges {
+    let actual = target.difference(from: source).inferringMoves()
+    expectEqual(actual, CollectionDifference(expected), "failed test at line \(line)")
```
StdlibUnittest `expect` functions take expected first and actual second.

moiseev commented on test/stdlib/Diffing.swift:
```diff
@@ -0,0 +1,667 @@
+  let mine = "Is\nit\nreview\ntime\nalready?"
+
+  // Split the contents of the sources into lines
+  let baseLines = base.components(separatedBy: "\n")
```
Isn't it an NSString API? How is it even available here without `import Foundation`, I wonder.

milseman replied:
Maybe through StdlibUnittest

numist replied:
I've only run the tests on Darwin, it seems likely that Foundation is getting imported somewhere. I'll make the test more literal.

moiseev commented on test/stdlib/Diffing.swift:
```diff
@@ -0,0 +1,667 @@
+  let expectedChanges: [(
+    source: [String],
+    target: [String],
+    changes: [CollectionDifference<String>.Change],
```
Oh.. So we don't need availability checks if the APIs are marked as 9999?

lorentey replied:
Ah, good catch -- I haven't run the tests after the availability commit.

milseman replied:
I still have to guard for it, but it will succeed at runtime

numist committed:
Incorporate feedback from @lorentey and @moiseev

lorentey committed:
[test] Add missing runtime availability check around diffing tests

numist committed:
Merge pull request #2 from lorentey/numist/diffing

[test] Add missing runtime availability check around diffing tests

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

numist committed:
Merge branch 'master' into numist/diffing

numist commented:
@swift-ci please test

Catfish-Man commented:
@swift-ci please test

lorentey commented:
The macOS failure in Swift Syntax was fixed in https://github.com/apple/swift-syntax/pull/106. 
Foundation's test_xpath failure on Linux also seems unrelated; it hasn't occurred in neighboring builds.

```
19:10:48 Test Case 'TestXMLDocument.test_xpath' started at 2019-03-06 01:10:47.995
19:10:48 
19:10:48 
19:10:48 0% tests passed, 1 tests failed out of 1
19:10:48 
19:10:48 Total Test time (real) =  21.82 sec
19:10:48 
19:10:48 The following tests FAILED:
19:10:48 	  1 - TestFoundation (SEGFAULT)
19:10:48 Errors while running CTest
19:10:48 FAILED: CMakeFiles/test.util 
```

lorentey commented:
Third time the charm!

@swift-ci test

numist commented:
@lorentey how did u do that
